### DISCUSSION TITLE
Batch text processing

### DISCUSSION THREAD

**Discussion by Definelymes** [urchade/GLiNER#73] [Ideas]
Hi, I was wondering if there is a feature (or would be in future) to process multiple texts at a same time instead of processing them one by one in sequential manner.

**Comment by urchade** [comment]
You can do it like this:

```python
def batch_predict_entities(self, texts, labels, flat_ner=True, threshold=0.5):
        """
        Predict entities for a batch of texts.
        texts:  List of texts | List[str]
        labels: List of labels | List[str]
        ...
        """
```

**Comment by Definelymes** [comment]
Thank you very much, I tried this approach and came accross this issue:
https://github.com/urchade/GLiNER/issues/63

Is there any plan to resolve it anytime soon on the new releases?

**Comment by urchade** [comment]
Fixed now. You can install v0.1.9 @Definelymes

**Comment by Hitish99** [comment]
Hi @urchade Urchade 
       Can you please share that while batch inference text are processed in parallel or concurrent?

**Comment by urchade** [comment]
Hi, The batch processing are processed in parallel @Hitish99

**Comment by batmech** [comment]
How does this batch processing really works? 

```
cleaned_text_list = cleaned_text.split()

chunk_list = [' '.join(cleaned_text_list[i:i + 160]).strip() for i in range(0, len(cleaned_text_list), 160)]

for i in range(0, len(chunk_list), 8):

    chunk_batch = chunk_list[i:i + 8]

    batch_results = model.batch_predict_entities(chunk_batch, labels=labels)
```

Above is my implementation for self-study. 
My question is: Does batch_predict_entities run on different CPU cores to facilitate batch processing (in parallel)?
@urchade

**Comment by amit-plutoflume** [comment]
Hi,

I’m trying to perform batch inference using GLiNER and noticed that the batch method seems slower than the sequential one.

Here’s what I tried:

# Batch processing
model.batch_predict_entities(texts, labels, threshold=0.5)

# Sequential processing
df["text"].apply(lambda x: model.predict_entities(x, labels, threshold=0.5))


Surprisingly, the second (sequential) approach is faster in my case.
Could you please suggest whether I might be using the batch processing method incorrectly, or if there are any configuration parameters (e.g., batch size, device setup) I should adjust?

Thank you

**Comment by Shrinivas23** [comment]
I agree with folks above, batch processing is slower than sequential.  Making it cost prohibitive to use in production with high traffic sadly. @urchade

**Comment by vivekkalyanarangan30** [comment]
@Shrinivas23 sequence packing should solve this in v0.2.23, I pushed a PR for this and there is a ~2.5x speedup in workloads, benchmarks scripts contain different test distributions
